charmcraft/commands/store/__init__.py

```diff
@@ -104,11 +114,82 @@ def fill_parser(self, parser):
         parser.add_argument(
             "--export", type=pathlib.Path, help="The file to save the credentials to"
         )
+        parser.add_argument(
+            "--charm",
+            action="append",
+            help=(
+                "The charm(s) on which the required credentials would work "
+                "(this option can be indicated multiple times; defaults to all)"
+            ),
+        )
+        parser.add_argument(
+            "--bundle",
+            action="append",
+            help=(
+                "The bundle(s) on which the required credentials would work "
+                "(this option can be indicated multiple times; defaults to all)"
+            ),
+        )
+        parser.add_argument(
+            "--channel",
+            action="append",
+            help=(
+                "The channel(s) on which the required credentials would work "
+                "(this option can be indicated multiple times, defaults to any channel)"
+            ),
+        )
+        parser.add_argument(
+            "--permission",
+            action="append",
+            help=(
+                "The permission(s) that the required credentials will have "
+                "(this option can be indicated multiple times, defaults to all permissions)"
+            ),
+        )
+        parser.add_argument(
+            "--ttl",
+            type=int,
+            help=(
+                "The time-to-live (in seconds) of the required credentials (defaults to 30 hours)"
+            ),
+        )
 
     def run(self, parsed_args):
         """Run the command."""
+        # validate that restrictions are only used if credentials are exported
+        restrictive_options = ["charm", "bundle", "channel", "permission", "ttl"]
+        if any(getattr(parsed_args, option) is not None for option in restrictive_options):
+            if parsed_args.export is None:
+                # XXX Facundo 2021-11-17: This is imported here to break a cyclic import. It will
+                # go away when we move this error to craft-cli lib.
+                from charmcraft.main import ArgumentParsingError
+
+                raise ArgumentParsingError(
+                    "The restrictive options 'bundle', 'channel', 'charm', 'permission' or 'ttl' "
+                    "can only be used when credentials are exported."
+                )
+        if parsed_args.permission is not None:
+            invalid = set(parsed_args.permission) - VALID_ATTENUATIONS
+            if invalid:
+                invalid_text = ", ".join(map(repr, sorted(invalid)))
+                raise CommandError(f"Invalid permission: {invalid_text}.")
+
+        # restrictive options, mapping the names between what is used in Namespace (singular,
+        # even if it ends up being a list) and the more natural ones used in the Store layer
```

Comment:
I had to read this comment a few times to understand it, then it became very obvious (to the point I would consider to removing it altogether).

Reply:
I like it, at some point explains why a "translation" is needed, we can improve it later.